### DISCUSSION TITLE
Should we "propagate const-ness" when doing structured binding?

### DISCUSSION THREAD

**Discussion by lanyizi** [nlohmann/json#3353] [Ideas]
Currently this surprisingly compiles:
https://godbolt.org/z/Ees7KfxzP

I don't have much knowledge on structured binding, but I guess it's related to the return type of this function?
https://github.com/nlohmann/json/blob/4f8fba14066156b73f1189a2b8bd568bde5284c5/include/nlohmann/detail/iterators/iteration_proxy.hpp#L156

**Comment by gregmarr** [comment]
The `const` here refers to the temporary that is implicitly constructed by the structured bindings.

    for (auto const& [key, value] : json.items()) {
        value.at(1) = 5;
    }

This is basically equivalent to this, which also compiles:

    for (auto const& item : json.items()) {
        auto &key = item.key();
        auto &value = item.value();
        value.at(1) = 5;
    }
